% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/methodCB.R
\name{methodCB}
\alias{methodCB}
\title{Adjust association statistics for collider bias and weak instrument bias}
\usage{
methodCB(
  xbeta,
  xse,
  ybeta,
  yse,
  prune = NULL,
  method,
  weighted = "1",
  model = "ivw",
  B = 100,
  lambda = seq(0.25, 5, 0.25),
  od = FALSE,
  seed = 0
)
}
\arguments{
\item{xbeta}{Vector of effects on the index trait.}

\item{xse}{Vector of standard errors of \code{xbeta}.}

\item{ybeta}{Vector of effects on the subsequent trait.}

\item{yse}{Vector of standard errors of \code{ybeta}.}

\item{prune}{Vector containing the indices of an approximately independent subset of the predictors in \code{xbeta} and \code{ybeta}. If unspecified, all predictors will be used.}

\item{method}{Method to adjust for regression dilution in the regression of \code{ybeta[prune]} on \code{xbeta[prune]}.
"CWLS" applies corrected weighted least squares estimator, "Fast" applies the fast version of CWLS, "SH" applies Slope Hunter and "mr.raps" applies MR-RAPS.}

\item{weighted}{Weights to be used in raw regression, Fast SIMEX and Hedges-Olkin adjustment. Default is using "1" as the first order weight, while "0" indicates unweighted model and "2" applies second order weight.}

\item{model}{Rgression model to be used in raw regression, Fast SIMEX and Hedges-Olkin adjustment. Default is "IVW" and could be replaced by "Egger" to use MR-Egger model.}

\item{B}{Number of simulations performed in each stage of the Fast SIMEX adjustment.}

\item{lambda}{Vector of lambdas for which the Fast SIMEX simulations are performed.}

\item{od}{Whether to use "over.dispersion" in MR-RAPS.}

\item{seed}{Random number seed for the Fast SIMEX adjustment.}
}
\value{
An object of class "methodCB" which contains:
  \itemize{
    \item{\code{ybeta.adj} Adjusted effects on the subsequent trait}
    \item{\code{yse.adj} Adjusted standard errors of \code{ybeta.adj}}
    \item{\code{ychisq.adj} Chi-square statistics for \code{(ybeta.adj/yse.adj)^2}}
    \item{\code{yp.adj} P-values for \code{ychisq.adj} on 1df}
    \item{\code{b} Coefficient of the regression of \code{ybeta[prune]} on \code{xbeta[prune]}, after correction for regression dilution}
    \item{\code{b.se} Standard error of \code{b}}
    \item{\code{b.ci} Lower and upper confidence limits for \code{b}}
    \item{\code{b.raw} Regression coefficient without correction for regression dilution}
    \item{\code{b.raw.se} Standard error of \code{b.raw}}
  }
}
\description{
Given effect sizes and standard errors for predictors of an index trait and a subsequent trait,
this function adjusts the statistics for the subsequent trait for selection bias through the index trait.
}
\details{
Effect sizes are on a linear scale, so could be the coefficients from linear regression, or log odds ratios, or log hazard ratios.
Effects on the subsequent trait are regressed on the effects on the index trait.
By default, the regression is weighted by the inverse variances of the subsequent trait effects.
The regression is adjusted for sampling variation in the index trait effects,
and the residuals then used to obtain adjusted effect sizes and standard errors for the subsequent trait.

The regression should be performed on a subset of predictors that are independent.
In the context of a genome-wide association study, these would be LD-pruned SNPs.
In terms of the input parameters, the regression command is \code{lm(ybeta[prune]~xbeta[prune],weights=1/yse[prune]^2)}.

The effects in \code{xbeta} and \code{ybeta} should be aligned for the same variables
and the same direction prior to running \code{indexevent}.

The default value of \code{B} is 100 to get a quick result, but higher values are recommended, eg 10000.
}
\references{
Cai S, Hartley A, Mahmoud O, Tilling K, Dudbridge F. Adjusting for collider bias in genetic association studies using instrumental variable methods. Genet Epidemiol. 2022;46:303â€“16.

Mahmoud O, Dudbridge F, Smith G D, et al. Slope-Hunter: A robust method for index-event bias correction in genome-wide association studies of subsequent traits[J]. bioRxiv, 2020.

Zhao, Q, et al. "Statistical inference in two-sample summary-data Mendelian randomization using robust adjusted profile score." The Annals of Statistics 48.3 (2020): 1742-1769.
}
\author{
Siyang Cai, Frank Dudbridge
}
